bookserver/crud.py

```diff
@@ -118,10 +127,118 @@ async def fetch_course(course_name: str) -> CoursesValidator:
     return CoursesValidator.from_orm(res.scalars().first())
 
 
+async def create_course(course_info: CoursesValidator) -> CoursesValidator:
+    new_course = Courses(**course_info.dict())
+    async with async_session() as session:
```

Comment:
Why not async_session.begin(), then omit the commit()?

Reply:
wouldn't it have to be `async_session().begin()` ??

Reply:
Actually I think they have to be nested in order for you to have a session variable.  In that case it is just a matter of style.

```
async with async_session() as session:
    async with session.begin():
        session.execute(....)
```

versus

```
async with async_session() as session:
    session.execute(...)
    await session.commit()
```

Reply:
I poked around -- the magic "one-liner" syntax is:

``` Python
async with async_session() as session, session.begin():
```

We should use this syntax, since it implies a rollback if an exception is raised as well as a commit on success. See https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block.

Reply:
NICE!

Reply:
OK, and also to your worries about changes to objects in the session, see: https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.close -- that seems to say that nad the end of the with block session.close() is called.

Reply:
I read a bit more -- since we're using a `sessionmaker`, even [better syntax](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#using-a-sessionmaker) is `with Session.begin() as session` -- see below (copied from the docs):

``` Python
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

# an Engine, which the Session will use for connection
# resources
engine = create_engine('postgresql://scott:tiger@localhost/')

# a sessionmaker(), also in the same scope as the engine
Session = sessionmaker(engine)

# we can now construct a Session() and include begin()/commit()/rollback()
# at once
with Session.begin() as session:
    session.add(some_object)
    session.add(some_other_object)
# commits the transaction, closes the session
```

The sqlalchemy docs do indicate that a `session.close()` (automatically performed by the `with` context manager) will detach everything from the session, which is exactly what we want. So, we can return results from `crud.py` as the original SQLAlchemy ORM classes if we want, instead of wrapping them in a Pydantic class.